### PR TITLE
Design & Implement Multi Level Neighbor Reduce

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

This is the PR to design, implement & test the MLNR concept as encountered in ICON.

- [x] Design the [concept](https://docs.google.com/document/d/1jA4lojFLrJEssYTzUT-o_0auobLqU3fnQB11YjFkaiE/edit?usp=sharing).
- [x] Adapt `atlas_interface.hpp`, `toylib_interface.hpp` and `unstructured_interface.hpp` to support the MLNR (both neighbor retrieval and looping)
- [x] Ensure that the algorithm returning the neighbors does so in the partial ordering described in the document above
- [x] Adapt the `ReductionOverNeighborExpr` to accept a chain of neighbors instead of a single right and left hand side, including serialization
- [x] Ensure that the `ReductionOverNeighborExpr` only ever accepts chains following the new notation (e.g. C->C becomes C->E->C). 
- [x] Introduce tests testing both the interfaces as well as the reduction. 

## Resolves / Enhances
Fixes https://github.com/MeteoSwiss-APN/dawn/issues/853
Fixes https://github.com/MeteoSwiss-APN/dawn/issues/854
Fixes https://github.com/MeteoSwiss-APN/dawn/issues/856

### CODE REVIEW

File: dawn/src/interface/atlas_interface.hpp [hunk 1]
```diff
@@ -192,6 +199,183 @@ std::vector<int> const nodeNeighboursOfNode(atlas::Mesh const& m, int const& idx
...
+      hybridTable_;
+
+public:
+  ConnInterface(const atlas::mesh::HybridElements::Connectivity& nbhTable)
+      : hybridTable_(nbhTable) {
+    connType_ = ConnType::Hybrid;
+  };
+  ConnInterface(const atlas::mesh::Nodes::Connectivity& nbhTable) : nodesTable_(nbhTable) {
+    connType_ = ConnType::Nodes;
+  };
```
[the remark this answers is not in the document]
Comment (Stagno) [reply dawn/src/interface/atlas_interface.hpp]: altough it seems that polymorphism is a better solution here (deriving ConnInterface for nodes and hybridelements)

[the remark this answers is not in the document]
Comment (Stagno) [reply dawn/src/interface/atlas_interface.hpp]: another idea: if atlas::mesh::Nodes::Connectivity and atlas::mesh::HybridElements::Connectivity share a common base class with the same interface as the one we want, we should probably use that (it seems from the usage below that they do)

File: dawn/src/interface/atlas_interface.hpp [hunk 2]
```diff
@@ -192,6 +199,183 @@ std::vector<int> const nodeNeighboursOfNode(atlas::Mesh const& m, int const& idx
...
+    std::copy(newFront.begin(), newFront.end(), std::inserter(result, result.end()));
+  }
+
+  if(chain.size() >= 2) {
+    getNeighborsImpl(nbhTables, chain, targetType, newFront, result);
+  }
+}
+
+// entry point, kicks off the recursive function above if required
+std::vector<int> getNeighbors(atlas::Mesh const& mesh, std::vector<dawn::LocationType> chain,
```
Comment (Stagno) [inline dawn/src/interface/atlas_interface.hpp]: Maybe it should return `std::set<int>`

File: dawn/src/interface/atlas_interface.hpp [hunk 3]
```diff
@@ -192,6 +199,183 @@ std::vector<int> const nodeNeighboursOfNode(atlas::Mesh const& m, int const& idx
...
+  // if next element is of target type we collect the current front into the result
+  if(to == targetType) {
+    std::copy(front.begin(), front.end(), std::inserter(result, result.end()));
+  }
+
+  // if there are two or more elements in the chain remaining, we need to recursively keep
+  // collecting neighbors
+  if(chain.size() >= 2) {
+    getNeighborsImpl(nbhTables, chain, targetType, front, result);
+  }
```
Comment (Stagno) [inline dawn/src/interface/atlas_interface.hpp]: you can avoid the recursion and the two `if(chain.size() >= 2)` if you use `getNeighborsImpl()` as a function that adds to the result set the "selected" neighbors on a single link of the chain (e.g. middle C->V in E->**C->V-**>E->V) and you loop (here) through the chain

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

**Comment by mroethlin** [comment]
launch jenkins

File: dawn/test/unit-test/interface/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+##
+##===------------------------------------------------------------------------------------------===##
+
+include(GoogleTest)
+
+if(DAWN_REQUIRE_UNSTRUCTURED_TESTING)
+  find_package(eckit REQUIRED)
+  find_package(atlas REQUIRED)
+else()
+  find_package(eckit QUIET)
```
[the remark this answers is not in the document]
Comment (Stagno) [reply dawn/test/unit-test/interface/CMakeLists.txt]: Same as CMakeLists.txt of unstructured integration tests, which was last refactored by @jdahm. My understanding is that if `DAWN_REQUIRE_UNSTRUCTURED_TESTING` is not set it should still look for eckit + atlas, but not complain if they are not found. Tests are then run if both packages are found.

[the remark this answers is not in the document]
Comment (twicki) [reply dawn/test/unit-test/interface/CMakeLists.txt]: :+1:

File: dawn/src/dawn/Serialization/SIRSerializer.cpp [hunk 5]
```diff
@@ -23,6 +23,7 @@
 #include "dawn/Support/Format.h"
 #include "dawn/Support/Logging.h"
 #include "dawn/Support/Unreachable.h"
+#include "driver-includes/unstructured_interface.hpp"
```
[the remark this answers is not in the document]
Comment (mroethlin) [reply dawn/src/dawn/Serialization/SIRSerializer.cpp]: include was accidental, the enum is in fact replicated on both end (`ast::LocationType` vs `dawn::LocationType`)

**Comment by mroethlin** [comment]
launch jenkins

**Pull request merged**